pkg/sdk/sdk.go

```diff
@@ -0,0 +1,147 @@
+package sdk
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/PereRohit/util/log"
+	respModel "github.com/PereRohit/util/model"
+	"github.com/gorilla/mux"
+	"io"
+	"io/ioutil"
+	"mime/multipart"
+	"net/http"
+	"time"
+)
+
+type htmlToPdfSvc struct {
+	svcUrl string
+}
+
+func NewHtmlToPdfSvc(url string) HtmlToPdfSvcI {
+	return &htmlToPdfSvc{
+		svcUrl: url,
+	}
+}
+
+type HtmlToPdfSvcI interface {
+	Register([]byte) (string, error)
+	Replace([]byte, string) error
+	GeneratePdf(map[string]interface{}, string) ([]byte, error)
+}
+
+func (h *htmlToPdfSvc) Register(fileBytes []byte) (string, error) {
+	//take in the slice of bytes so that you can pass it directly to io.copy
+	//take io.reader as argument
+	file := bytes.NewReader(fileBytes)
+	body := &bytes.Buffer{}
+	writer := multipart.NewWriter(body)
+	part, err := writer.CreateFormFile("file", "output")
+	if err != nil {
+		return "", err
+	}
+	_, err = io.Copy(part, file)
+	writer.Close()
+	client := http.Client{
+		Timeout: 5 * time.Second,
+	}
+
+	r, err := client.Post(h.svcUrl+"/v1/register", writer.FormDataContentType(), body)
+	if err != nil {
+		return "", err
+	}
+
+	if r.StatusCode < 200 || r.StatusCode > 299 {
+		return "", fmt.Errorf("non success status code received : %v", r.StatusCode)
+	}
+	resp, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return "", err
+	}
+	var response respModel.Response
+	err = json.Unmarshal(resp, &response)
+	if err != nil {
+		return "", err
+	}
+	x := response.Data
+	m, ok := x.(map[string]interface{})
+	if !ok {
+		errNew := errors.New("unable to parse response data")
+		return "", errNew
+	}
+	return fmt.Sprint(m["id"]), err
+}
+
+func (h *htmlToPdfSvc) Replace(fileBytes []byte, id string) error {
```

Comment:
All the below test scenarios need to be covered:

1. New request failure
2. HTTP call failure
3. non 200 status code received
4. ReadAll failure
5. Unmarshal failure
6. map inference failure
7. Success case

Reply:
I have included 6 test cases except readall failure

Reply:
cannot fail the readall as server doesnot allow response body as nil

Reply:
Cases 1 & 2 are yet to be covered.

Reply:
Which test cases points to which scenario for 1 & 2?

Reply:
new request failure not possible